src/tools/upgrade-factory-snapshot.c

```diff
@@ -0,0 +1,725 @@
+        partition_total = 1;
+    }
+
+    ufs->status.partition_current = partition_current;
+    ufs->status.partition_total = partition_total;
+
+    SFMF_LOG("queue: %s, task='%s' (%d/%d), partition='%s' (%d/%d), message='%s' (%d%%)\n",
+            queue->name, task->name, queue->current+1, queue->total,
+            partition, partition_current+1, partition_total,
+            message, progress);
+
+    GError *error = NULL;
+    if (!g_dbus_connection_emit_signal(ufs->system_bus, NULL, UFS_DBUS_PATH, UFS_DBUS_INTERFACE, "ProgressChanged",
+                g_variant_new("()"), &error)) {
+        SFMF_WARN("Could not forward progress via D-Bus: %s\n", error->message);
+        g_error_free(error);
+    }
+}
+
+gboolean upgrade_factory_snapshot_idle_timeout_cb(gpointer user_data)
+{
+    struct UpgradeFactorySnapshot *ufs = user_data;
+
+    if (!ufs->running) {
+        SFMF_DEBUG("Idle timeout reached, quitting\n");
+        g_main_loop_quit(ufs->mainloop);
+    }
+
+    return FALSE;
+}
+
+void upgrade_factory_snapshot_schedule_quit(struct UpgradeFactorySnapshot *ufs)
+{
+    if (ufs->idle_timer_source_id) {
+        g_source_remove(ufs->idle_timer_source_id);
+        ufs->idle_timer_source_id = 0;
+    }
+
+    SFMF_DEBUG("(Re-)Starting idle timer (%d seconds)\n", IDLE_TIMEOUT_SEC);
+    ufs->idle_timer_source_id = g_timeout_add_seconds(IDLE_TIMEOUT_SEC,
+            upgrade_factory_snapshot_idle_timeout_cb, ufs);
+}
+
+void upgrade_factory_snapshot_set_running(struct UpgradeFactorySnapshot *ufs, gboolean running)
+{
+    if (ufs->running != running) {
+        GError *error = NULL;
+
+        GVariantBuilder builder;
+        g_variant_builder_init(&builder, G_VARIANT_TYPE_ARRAY);
+        g_variant_builder_add(&builder, "{sv}", "running", g_variant_new_boolean(running));
+
+        if (!g_dbus_connection_emit_signal(ufs->system_bus, NULL, UFS_DBUS_PATH,
+                    "org.freedesktop.DBus.Properties", "PropertiesChanged",
+                    g_variant_new("(sa{sv}as)", UFS_DBUS_INTERFACE, &builder, NULL), &error)) {
+            SFMF_WARN("Could not emit properties changed signal: %s\n", error->message);
+            g_error_free(error);
+        }
+
+        ufs->running = running;
+    }
+}
+
+void upgrade_factory_snapshot_dbus_signal_cb(GDBusConnection *connection,
+        const gchar *sender_name, const gchar *object_path, const gchar *interface_name,
+        const gchar *signal_name, GVariant *parameters, gpointer user_data)
+{
+    struct UpgradeFactorySnapshot *ufs = user_data;
+
+    gchar *args = g_variant_print(parameters, TRUE);
+    //SFMF_DEBUG("dbus: sender=%s path=%s signal=%s.%s%s\n", sender_name, object_path,
+    //        interface_name, signal_name, args);
+    g_free(args);
+
+    if (strcmp(signal_name, "Progress") == 0) {
+        gchar *partition;
+        int progress;
+        gchar *message;
+        g_variant_get(parameters, "(sis)", &partition, &progress, &message);
```

Comment:
Is it guaranteed that partition & message get a value? Would feel a lot safer if at least the pointers were initialized.